### PR TITLE
MRDC-722

### PR DESCRIPTION
**Pull request by clample**
Client changes for multi DC Appium suites

### CODE REVIEW

**Comment by clample** [comment]
To run on a TestObject instance other than production (staging, locally, etc.), you need to set the app API endpoint in two places:

* in the watcher: `watcher.setRemoteWebdriver(driver)` => `watcher.setRemoteWebDriver(driver, new URL("http://127.0.0.1:7070/api"))`
* as a environment variable: `TESTOBJECT_API_ENDPOINT=http://127.0.0.1:7070/api`

This was the case before these changes. Just clarifying here for others.

File: pom.xml [hunk 1]
```diff
@@ -140,7 +140,10 @@
         <dependency>
             <groupId>org.testobject</groupId>
             <artifactId>testobject-java-api</artifactId>
-            <version>0.0.26</version>
+            <version>0.0.27</version>
+            <!-- Using local jar for development. Remove when releasing -->
```
Comment (umutuzgur) [inline pom.xml]: Remove :D

Comment (umutuzgur) [inline pom.xml]: You can also just install the java-api locally with mvn install. You don't need to have a configuration like this

Comment (clample) [inline pom.xml]: Haha ;)

Could you give me more information on the mvn install trick (do you have a link or something I could google), that might be really helpful

File: pom.xml [hunk 2]
```diff
@@ -6,7 +6,7 @@
 
     <groupId>org.testobject</groupId>
     <artifactId>testobject-appium-java-api</artifactId>
-    <version>0.0.27</version>
+    <version>0.0.28</version>
```
Comment (umutuzgur) [inline pom.xml]: We should do the same thing here

File: src/main/java/org/testobject/appium/junit/TestObjectAppiumSuite.java [hunk 3]
```diff
@@ -211,20 +200,26 @@ private static TestObject getConfig(Class<?> clazz) {
...
 
-	private Set<String> getRemoteDeviceIds() {
+	private Set<DataCenterSuite> getDataCenterSuites() {
 		AppiumSuiteResource suiteReportResource = new AppiumSuiteResource(client);
-
-		return suiteReportResource.readSuiteDeviceIds(testObjectSuiteId);
+		Set<DataCenterSuite> dataCenterSuites = suiteReportResource.readDataCenterSuites(testObjectSuiteId);
+		for (DataCenterSuite dataCenterSuite : dataCenterSuites) {
+			if (dataCenterSuite.getDeviceDescriptorIds().isEmpty() == false) {
+				return dataCenterSuites;
```
Comment (umutuzgur) [inline src/main/java/org/testobject/appium/junit/TestObjectAppiumSuite.java]: I don't quite get this logic. What did you try to do here? If you are doing validation here maybe putting that into a method to explain it would be a better idea

File: src/main/java/org/testobject/appium/junit/TestParser.java [hunk 4]
```diff
@@ -8,39 +8,14 @@
...
-
-
 	public static Test from(Description testDescription) {
+		String className = testDescription.getClassName();
 
-		String className;
-		String methodName;
-		String deviceId = null;
-
-		Matcher matcher = METHOD_AND_CLASS_NAME_PATTERN_STRICT.matcher(testDescription.getMethodName());
```
Comment (umutuzgur) [inline src/main/java/org/testobject/appium/junit/TestParser.java]: Soo, this was unnecessary?

File: src/test/java/org/testobject/appium/junit/RemoteWebDriverCalculatorSuiteTestJUnit.java [hunk 5]
```diff
@@ -33,7 +33,7 @@ public void setup() throws MalformedURLException {
         capabilities.setCapability(TestObjectCapabilities.TESTOBJECT_API_KEY, watcher.getApiKey());
         capabilities.setCapability(TestObjectCapabilities.TESTOBJECT_TEST_REPORT_ID, watcher.getTestReportId());
 
-        driver = new RemoteWebDriver(TestObjectCapabilities.TESTOBJECT_APPIUM_ENDPOINT, capabilities);
+        driver = new RemoteWebDriver(watcher.getTestObjectOrLocalAppiumEndpointURL(), capabilities);
```
Comment (umutuzgur) [inline src/test/java/org/testobject/appium/junit/RemoteWebDriverCalculatorSuiteTestJUnit.java]: Can't we just turn this into getAppiumEndpointURL and put java doc? It is a really ugly looking method name. Maybe you should ask this to Andreas probably

File: src/main/java/org/testobject/appium/junit/TestParser.java (hunk 4 above)
Comment (clample) [inline src/main/java/org/testobject/appium/junit/TestParser.java]: We still parse the test, we just now make sure the test is in a much simpler format (everything separated by spaces). 

I think the old parsing was very complicated, so the change makes sense.

Comment (clample) [inline src/main/java/org/testobject/appium/junit/TestParser.java]: The method name, device id, and data center id should never have spaces, so this should work fine.

File: src/test/java/org/testobject/appium/junit/RemoteWebDriverCalculatorSuiteTestJUnit.java (hunk 5 above)
Comment (clample) [inline src/test/java/org/testobject/appium/junit/RemoteWebDriverCalculatorSuiteTestJUnit.java]: Changing the method name will require users to change their test code. I agree it's ugly, though.

I think adding javadoc for the method or changing it's name should be in a separate PR, since it's not really related to these changes.

File: src/main/java/org/testobject/appium/junit/TestObjectAppiumSuite.java (hunk 3 above)
Comment (clample) [inline src/main/java/org/testobject/appium/junit/TestObjectAppiumSuite.java]: Good catch. I agree, this doesn't make sense. I changed the behaviour in https://github.com/testobject/testobject-appium-java-api/pull/18/commits/45836cba1a9589632d45357a1c31f1c80d9f964a

**Comment by clample** [comment]
Working on staging: https://app.staging.testobject.org/#/foobar/calculator/appium/4426/reports/2210

**Comment by clample** [comment]
travis is failing because testobject-java-api:0.1.0 isn't available on maven central yet. This is an expected error.
